### PR TITLE
Improved performance of loading embedding matrix

### PR DESCRIPTION
**Pull request by qubixes**
Use Queue's from the multiprocessing library to speed up the loading times of the embedding layer. Also, use the knowledge of the sampling to not having to load the whole matrix into memory.

On a MacBook Pro Retina 2012 there is a 10x speed up, now loading in approximately 30s.

### CODE REVIEW

**Comment by J535D165** [comment]
Nice PR and nice speed up.

Is there a reason to choose queus over a pool?

**Comment by qubixes** [comment]
I'm no expert on either, but Queue's seem more convenient with a simultaneous IO/worker process configuration? Pool's seem more useful when all the data is already present.

File: asr/models/embedding.py [hunk 1]
```diff
@@ -1,9 +1,88 @@
-"""Script to create the memory of an embeeding layer."""
+'''
+Created on 22 Feb 2019
```
Comment (J535D165) [inline asr/models/embedding.py]: Remove

File: asr/models/embedding.py [hunk 2]
```diff
@@ -1,9 +1,88 @@
...
+
+@author: Raoul Schram, Parisa Zahedi, Jonathan de Bruin
```
Comment (J535D165) [inline asr/models/embedding.py]: remove

File: asr/models/embedding.py [hunk 3]
```diff
@@ -1,9 +1,88 @@
...
+
+Multiprocessing version of embedding file. 
+'''
 
 import numpy as np
 
+from multiprocessing import Process, Queue
```
Comment (J535D165) [inline asr/models/embedding.py]: CPython imports before external imports. PEP8

File: asr/models/embedding.py [hunk 4]
```diff
@@ -1,9 +1,88 @@
...
+
+def _embedding_reader(filename, input_queue, block_size=1):
```
Comment (J535D165) [inline asr/models/embedding.py]: default block size 1000, remove in #L108 and #L134.

File: asr/models/embedding.py [hunk 5]
```diff
@@ -1,9 +1,88 @@
...
+    while True:
+        buffer = input_queue.get()
+        if buffer == "DONE":
+            break
+        
+        for line in buffer:
+            values = line.split(' ')
+            
+            if len(values) != d + 2:
+                raise ValueError("Check values: %s" % values)
```
Comment (J535D165) [inline asr/models/embedding.py]: Use `f"Check values: {values}"`

File: asr/models/embedding.py [hunk 6]
```diff
@@ -1,9 +1,88 @@
...
+            embedding_dict[word] = np.asarray(coefs, dtype=np.float32)
 
-def load_embedding(fp, verbose=1):
+    #We removed the "DONE" from the input queue, so put it back in for the other processes.
+    input_queue.put("DONE")
+    #Store the results in the output queue
+    output_queue.put(embedding_dict)
+    
+
+def load_embedding(fp, verbose=1, word_index=None):
```
Comment (J535D165) [inline asr/models/embedding.py]: Make `verbose` third argument.

File: asr/models/embedding.py [hunk 7]
```diff
@@ -15,42 +94,58 @@ def load_embedding(fp, verbose=1):
...
     dict:
         The embedding weights stored in a dict with the word as key and
         the weights as values.
     """
-
-    embedding = {}
-
+    
+    #Granularity of the worker jobs.
+    block_size=1000
```
Comment (J535D165) [inline asr/models/embedding.py]: remove

File: asr/models/embedding.py [hunk 8]
```diff
@@ -15,42 +94,58 @@ def load_embedding(fp, verbose=1):
...
+    #Maximum number of jobs in the queue.
+    queue_size=500
+    #Number of worker jobs running in parallel.
+    n_worker=3
```
Comment (J535D165) [inline asr/models/embedding.py]: Make this an argument named `n_jobs` (sklearn). Default 1. Value -1 should be all cores.

File: asr/models/embedding.py [hunk 9]
```diff
@@ -15,42 +94,58 @@ def load_embedding(fp, verbose=1):
...
         the weights as values.
     """
-
-    embedding = {}
-
+    
+    #Granularity of the worker jobs.
+    block_size=1000
+    #Maximum number of jobs in the queue.
+    queue_size=500
```
Comment (J535D165) [inline asr/models/embedding.py]: Hardcoding looks fine here

File: asr/models/embedding.py [hunk 10]
```diff
@@ -15,42 +94,58 @@ def load_embedding(fp, verbose=1):
...
-            coefs = values[1:d + 1]
-
-            # store the results
-            embedding[word] = np.asarray(coefs, dtype=np.float32)
-
+    embedding = {}
+    
+    worker_procs=[]
+    for _ in range(n_worker):
+        p = Process(target=_embedding_worker,
```
Comment (J535D165) [inline asr/models/embedding.py]: outline

Comment (J535D165) [reply asr/models/embedding.py]: We sometimes use yapf.

File: asr/models/embedding.py [hunk 11]
```diff
@@ -1,9 +1,88 @@
...
+    block_size: int
+        Number of lines for each job.
+    """
+    
+    with open(filename, 'r', encoding='utf-8', newline='\n') as f:
+        f.readline()
+        
+        ib=0
+        buffer=[]
+        for _, line in enumerate(f): # This can be done better probably...
```
Comment (J535D165) [inline asr/models/embedding.py]: remove comment

File: asr/models/embedding.py [hunk 12]
```diff
@@ -1,9 +1,88 @@
...
+    filename: str
+        File of trained embedding vectors.
+    input_queue: Queue
+        Queue to store jobs in.
+    block_size: int
+        Number of lines for each job.
+    """
+    
+    with open(filename, 'r', encoding='utf-8', newline='\n') as f:
+        f.readline()
```
Comment (J535D165) [inline asr/models/embedding.py]: Your aren't using the info. You can use this to improve for loop.

File: asr/models/embedding.py (hunk 11 above)
Comment (J535D165) [reply asr/models/embedding.py]: While loop with `try except` is what you are looking for.

File: asr/models/embedding.py [hunk 13]
```diff
@@ -1,9 +1,88 @@
...
+def _embedding_worker(input_queue, output_queue, d, word_index=None):
+    """ Process that reads the word embeddings from a file.
+    
+    Parameters
+    ----------
+    input_queue: Queue
+        Queue in which the jobs are submitted.
+    output_queue: Queue
+        Queue to store the embedding in dictionary form.
+    d: int
```
Comment (J535D165) [inline asr/models/embedding.py]: Explicit variables names `import zen`

File: asr/review.py [hunk 14]
```diff
@@ -43,8 +43,8 @@ def _load_embedding_matrix(fp, word_index):
 
     from asr.models.embedding import load_embedding, sample_embedding
 
-    embedding, words = load_embedding(fp)
```
Comment (J535D165) [inline asr/review.py]: good catch

**Comment by J535D165** [comment]
Shall we add a unittest before merging?

**Comment by J535D165** [comment]
And a function to download the embedding?

**Comment by qubixes** [comment]
I added unit tests, including testing for the URL you gave me. (Obviously it all passes.) 

I think all the things you mentioned are resolved.

One possible issue: if one of the processes gets an exception, the others will deadlock. Obviously, the error will still be raised. Solution would be a try/except block that only raises an error after everything will be joined. But of course the whole thing would be one more indented to the right...

**Comment by J535D165** [comment]
Please rebase.

**Comment by J535D165** [comment]
Can you add the following line to the __init__.py of the module? 

```
from asr.models.embedding import load_embedding, sample_embedding
```

**Comment by qubixes** [comment]
Should be able to merge now?

**Comment by J535D165** [comment]
What time should I think of? It takes ages (or hangs) on my machine. 

``` python
In [1]: import asr                                                                                                 

In [2]: %timeit asr.load_embedding("/Users/jonathan/asr_data/fasttext.cc.en.300.vec")                              
Reading 2000000 vectors with 300 dimensions.
```

**Comment by qubixes** [comment]
Hmmm, with 1 core and no sampling it should be similar to what it was before, so 5 min perhaps (depending on hardware obviously)?

On 12 Mar 2019, at 16:30, Jonathan de Bruin <notifications@github.com<mailto:notifications@github.com>> wrote:


What time should I think of? It takes ages (or hangs) on my machine.

In [1]: import asr

In [2]: %timeit asr.load_embedding("/Users/jonathan/asr_data/fasttext.cc.en.300.vec")
Reading 2000000 vectors with 300 dimensions.

**Pull request merged**